File: android/app/src/main/java/com/ccoan/ny/data/models/Schedule.kt
package com.ccoan.ny.data.models

import com.squareup.moshi.JsonClass

@JsonClass(generateAdapter = true)
data class ScheduleSlot(
    var phase: String,
    var department: String,
    var assignees: String
)

@JsonClass(generateAdapter = true)
data class SundaySchedule(
    val id: String? = null,
    val date: String,
    var slots: List<ScheduleSlot> = emptyList(),
    val createdBy: String? = null,
    val lastModifiedBy: String? = null,
    val metadata: MetadataTimestamps? = null
)

@JsonClass(generateAdapter = true)
data class ScheduleConfig(
    var departments: List<String>,
    var phases: List<String>
)

@JsonClass(generateAdapter = true)
data class ScheduleResponse(
    val success: Boolean,
    val data: SundaySchedule? = null,
    val canEdit: Boolean = false,
    val scheduleAdminId: String? = null,
    val scheduleAdminName: String? = null
)

@JsonClass(generateAdapter = true)
data class UpdateScheduleRequest(
    val date: String,
    val slots: List<ScheduleSlot>
)

@JsonClass(generateAdapter = true)
data class SetScheduleAdminRequest(
    val userId: String
)

object ScheduleDefaults {
    val DEFAULT_PHASES = listOf(
        "Before the Service",
        "Opening",
        "Worship",
        "Testimonies",
        "Sermon",
        "Prayer Line Intro",
        "Prayer Line",
        "Laying of Hands",
        "Mass Prayer",
        "Offering",
        "Grace",
        "After Service"
    )

    val DEFAULT_DEPARTMENTS = listOf(
        "LAUNDROMAT", "CAMERA", "Sound Dept.", "Media/Display", "SERMON",
        "SUNDAY SERVICE OPENING", "TRANSPORTATION", "PROTOCOL", "Photo Dept.",
        "Confirmation", "Office", "Testimony interview", "Accounting",
        "Microphone", "Translation", "Emergency", "Sunday School",
        "Usher", "Cleaning", "Worship"
    )
}

File: ios/CCOAN/Models/Schedule.swift
import Foundation

struct ScheduleSlot: Codable {
    var phase: String
    var department: String
    var assignees: String
}

struct SundaySchedule: Codable, Identifiable {
    let id: String?
    let date: String
    var slots: [ScheduleSlot]
    let createdBy: String?
    let lastModifiedBy: String?
    let metadata: MetadataTimestamps?

    enum CodingKeys: String, CodingKey {
        case id, date, slots, createdBy, lastModifiedBy, metadata
    }

    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        self.id = try container.decodeIfPresent(String.self, forKey: .id)
        self.date = try container.decode(String.self, forKey: .date)
        self.slots = try container.decodeIfPresent([ScheduleSlot].self, forKey: .slots) ?? []
        self.createdBy = try container.decodeIfPresent(String.self, forKey: .createdBy)
        self.lastModifiedBy = try container.decodeIfPresent(String.self, forKey: .lastModifiedBy)
        self.metadata = try container.decodeIfPresent(MetadataTimestamps.self, forKey: .metadata)
    }
}

struct ScheduleConfig: Codable {
    var departments: [String]
    var phases: [String]
}

struct ScheduleResponse: Decodable {
    let success: Bool
    let data: SundaySchedule?
    let canEdit: Bool
    let scheduleAdminId: String?
    let scheduleAdminName: String?
}

struct UpdateScheduleRequest: Encodable {
    let date: String
    let slots: [ScheduleSlot]
}

struct SetScheduleAdminRequest: Encodable {
    let userId: String
}
